tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1Closure.cpp

```diff
@@ -0,0 +1,101 @@
+                  "[Unit][M1Grey]") {
+  // Reference DataVector
+  const DataVector ref_dv(1);
+  // Create variables
+  // Input
+  Scalar<DataVector> energy_density(ref_dv);
+  tnsr::i<DataVector, 3, Frame::Inertial> momentum_density(ref_dv);
+  tnsr::I<DataVector, 3, Frame::Inertial> fluid_velocity(ref_dv);
+  Scalar<DataVector> fluid_lorentz_factor(ref_dv);
+  tnsr::ii<DataVector, 3, Frame::Inertial> spatial_metric(ref_dv);
+  // Output
+  Scalar<DataVector> closure_factor(ref_dv);
+  tnsr::II<DataVector, 3, Frame::Inertial> pressure_tensor(ref_dv);
+  Scalar<DataVector> comoving_energy_density(ref_dv);
+  Scalar<DataVector> comoving_momentum_density_normal(ref_dv);
+  tnsr::i<DataVector, 3, Frame::Inertial>
+      comoving_momentum_density_spatial(ref_dv);
+
+  // Allocate fluid/metric variables
+  for (size_t m = 0; m < 3; m++) {
+    fluid_velocity.get(m) = 0.1 * m;
+    spatial_metric.get(m, m) = 1. + 0.1 * m * m;
+    for (size_t n = m + 1; n < 3; n++) {
+      spatial_metric.get(m, n) = 0.1 * (m + n);
+    }
+  }
+  const auto det_and_inv = determinant_and_inverse(spatial_metric);
+  const auto& inv_spatial_metric = det_and_inv.second;
+  get(fluid_lorentz_factor) = 1.;
+  for (int m = 0; m < 3; m++) {
+    for (int n = 0; n < 3; n++) {
+      get(fluid_lorentz_factor) -= spatial_metric.get(m, n) *
+                                   fluid_velocity.get(m) *
+                                   fluid_velocity.get(n);
+    }
+  }
+  get(fluid_lorentz_factor) = 1. / sqrt(get(fluid_lorentz_factor));
+
+  // Initialize closure factor (as the input value is used as initial
+  // guess for the root finding algorithm).
+  get(closure_factor) = -1.;
+  // (1) Optically thick limit:
+  get(energy_density) = 1. / 3 * (4. * square(get(fluid_lorentz_factor)) - 1.);
+  for (int m = 0; m < 3; m++) {
+    momentum_density.get(m) = 0.;
+    for (int n = 0; n < 3; n++) {
+      momentum_density.get(m) += 4. / 3 * square(get(fluid_lorentz_factor)) *
+                                 fluid_velocity.get(n) *
+                                 spatial_metric.get(m, n);
+    }
+  }
+  RadiationTransport::M1Grey::M1Closure(
+      make_not_null(&closure_factor), make_not_null(&pressure_tensor),
+      make_not_null(&comoving_energy_density),
+      make_not_null(&comoving_momentum_density_normal),
+      make_not_null(&comoving_momentum_density_spatial), energy_density,
+      momentum_density, fluid_velocity, fluid_lorentz_factor, spatial_metric,
+      inv_spatial_metric);
+  CHECK(fabs(get(closure_factor)[0]) < 1.e-5);
+
+  // (2) Optically thin limit
+  momentum_density.get(0) = -1.;
+  momentum_density.get(1) = 5.;
+  momentum_density.get(2) = 3.;
+  get(energy_density) = 0.;
+  for (int m = 0; m < 3; m++) {
+    for (int n = 0; n < 3; n++) {
+      get(energy_density) += inv_spatial_metric.get(m, n) *
+                             momentum_density.get(m) * momentum_density.get(n);
+    }
+  }
+  get(energy_density) = sqrt(get(energy_density));
```

Comment:
There is an overload for `magnitude` (`src/DataStructures/Tensor/EagerMath/Magnitude.hpp`) that should save you the previous loop.